sherpa/sherpa
# Skip tests instead of pretending to pass

Tests are properly skipped when some conditions are not met, e.g. when there is no `FITS` backend available, or the external data/threads data folder is missing.

This is largely based on @taldcroft's original PR #10, but it removes the dependency on astropy (PR #6), it includes more checks, and covers the CIAO use cases, i.e. it also checks for the presence of `pycrates`.

To test:

```
python setup.py develop
sherpa_test
```
tests should be run with and without pyfits or pycrates, with and without xspec, with and without the test data directory (i.e. `sherpa_test -d /data/scialg/testdata`).

When running the full regression test suite (i.e. pyfits+testdata, with or without pyfits) some tests are expected to fail because of unmet dependencies (CIAO data model, pycrates), and they will be fixed in the PR that introduces the new testing infrastructure.

Below is a log of the tests that are expected to fail:
```
$ sherpa_test -d /data/scialg/testdata
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
Failed importing sherpa.astro.xspec: No module named _xspec
  Found 7/7 tests for sherpa.astro.tests.test_datastack
  Found 1/1 tests for sherpa.astro.models.tests.test_models
  Found 25/25 tests for sherpa.astro.tests.test_optical
  Found 2/2 tests for sherpa.astro.sim.tests.test_sim
  Found 26/26 tests for sherpa.astro.tests.test_astro
  Found 6/6 tests for sherpa.astro.tests.test_data
  Found 1/1 tests for sherpa.astro.tests.test_plot
  Found 1/1 tests for sherpa.astro.ui.tests.test_nan
  Found 15/15 tests for sherpa.astro.ui.tests.test_ui
  Found 1/1 tests for sherpa.astro.utils.tests.test_utils
  Found 4/4 tests for sherpa.estmethods.tests.test_estmethods
  Found 0/0 tests for sherpa.image.tests.test_image
  Found 1/1 tests for sherpa.models.tests.test_basic
  Found 14/14 tests for sherpa.models.tests.test_model
  Found 12/12 tests for sherpa.models.tests.test_parameter
  Found 6/6 tests for sherpa.models.tests.test_template
  Found 26/26 tests for sherpa.optmethods.tests.test_optmethods
  Found 18/18 tests for sherpa.plot.tests.test_plot
  Found 18/18 tests for sherpa.sim.tests.test_sim
  Found 5/5 tests for sherpa.tests.test_sherpa
  Found 1/1 tests for sherpa.tests.test_stack
  Found 8/8 tests for sherpa.tests.test_ui
  Found 1/1 tests for sherpa.utils.tests.test_err
  Found 1/1 tests for sherpa.utils.tests.test_integration
  Found 43/43 tests for sherpa.utils.tests.test_root
  Found 10/10 tests for sherpa.utils.tests.test_utils
.................................ss..s....E.s.sss....sss.ssss..........s..E..................................................................................................................................................................................
======================================================================
ERROR: test_lev3fft (sherpa.astro.tests.test_astro.test_threads)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/tests/test_astro.py", line 501, in test_lev3fft
    self.run_thread('lev3fft', scriptname='bar.py')
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/tests/test_astro.py", line 60, in run_thread
    execfile(scriptname, {}, self.locals)
  File "bar.py", line 15, in <module>
    notice2d_id(srcid, "region(" + reg_file + "[SRCREG])")
  File "<string>", line 1, in notice2d_id
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/ui/utils.py", line 6432, in notice2d_id
    self.get_data(id).notice2d(val, False)
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/data.py", line 2002, in notice2d
    os.path.isfile(val), ignore)
TypeError: unable to parse region string successfully

======================================================================
ERROR: test_chi2 (sherpa.astro.ui.tests.test_ui.test_stats_ui)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/ui/tests/test_ui.py", line 252, in test_chi2
    si=ui.get_stat_info()
  File "<string>", line 1, in get_stat_info
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/ui/utils.py", line 7660, in get_stat_info
    return self._get_stat_info()
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/ui/utils.py", line 9441, in _get_stat_info
    ids, datasets, models = self._prepare_fit(None)
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/ui/utils.py", line 7484, in _prepare_fit
    mod = self._get_model(i)
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/ui/utils.py", line 5552, in _get_model
    return self._add_convolution_models(id, data, model, is_source)
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/ui/utils.py", line 8325, in _add_convolution_models
    resp = sherpa.astro.instrument.Response1D(data)
  File "/data/scialg/staff/olaurino/working/sherpa/sherpa/astro/instrument.py", line 651, in __init__
    raise DataErr('norsp', pha.name)
DataErr: No instrument response found for dataset /data/scialg/testdata/threads/chi2/SWIFTJ0840.1+2946.pha.gz

----------------------------------------------------------------------
Ran 253 tests in 36.441s

FAILED (errors=2, skipped=15)
```
